[pkg] simplifications to broken dependency handling

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58573 [pkg] Catch exceptions where dependency resolution gets invalid imports
* **#58572 [pkg] simplifications to broken dependency handling**

Right now, we have three categories of error (broken, denied, unhandled). This
PR unifies them into a single "error" field in the node, with optional context.
It also generalizes how formatting of the error in PackagingError occurs.

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982/)

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982)

suo committed:
[pkg] simplifications to broken dependency handling

Right now, we have three categories of error (broken, denied, unhandled). This
PR unifies them into a single "error" field in the node, with optional context.
It also generalizes how formatting of the error in PackagingError occurs.

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982/)

[ghstack-poisoned]

suo committed:
Update on "[pkg] simplifications to broken dependency handling"


Right now, we have three categories of error (broken, denied, unhandled). This
PR unifies them into a single "error" field in the node, with optional context.
It also generalizes how formatting of the error in PackagingError occurs.

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982/)

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982)

[ghstack-poisoned]

SplitInfinity commented on test/package/test_dependency_api.py:
```diff
@@ -224,13 +224,9 @@ def test_intern_error(self):
         buffer = BytesIO()
 
-        try:
+        with self.assertRaisesRegex(PackagingError, "did not match against any action"):
```
This assertion is not exactly the same as the one that was removed. It would be good to also check that the packages listed as unhandled are the ones we expect.

SplitInfinity commented on test/package/test_dependency_api.py:
```diff
@@ -262,16 +258,12 @@ def import_module(self, module_name):
         buffer = BytesIO()
 
-        try:
+        with self.assertRaisesRegex(PackagingError, "C extension"):
```
Same here.

SplitInfinity replied:
Also this test fails for a different reason on Windows

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -82,61 +93,30 @@ class PackagingError(Exception):
+                continue
+            error = attrs["error"]
+            if error == PackagingErrorReason.NO_ACTION:
+                assert "action" not in attrs
```
Light reminder to make these attribute strings constants or something of that nature.

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -82,61 +93,30 @@ class PackagingError(Exception):
+            error = attrs["error"]
+            if error == PackagingErrorReason.NO_ACTION:
+                assert "action" not in attrs
+            broken.setdefault(error, []).append(module_name)
```
Also I still can't believe a method named `SETdefault` has a return value

SplitInfinity reviewed (approved):
:shipit:

suo committed:
Update on "[pkg] simplifications to broken dependency handling"


Right now, we have three categories of error (broken, denied, unhandled). This
PR unifies them into a single "error" field in the node, with optional context.
It also generalizes how formatting of the error in PackagingError occurs.

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982/)

Differential Revision: [D28543982](https://our.internmc.facebook.com/intern/diff/D28543982)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28543982).
